src/webui/api/torrentscontroller.cpp

```diff
@@ -1523,3 +1582,79 @@ void TorrentsController::setSSLParametersAction()
 
     torrent->setSSLParameters(sslParams);
 }
+
+void TorrentsController::metadataAction()
+{
+    const QString source {params()[u"source"_s]};
+
+    if (Net::DownloadManager::hasSupportedScheme(source))
+        throw APIError(APIErrorType::BadParams, tr("Source must be magnet URI or info hash"));
+
+    if (const auto parseResult = BitTorrent::TorrentDescriptor::parse(source))
+    {
+        auto *session = BitTorrent::Session::instance();
+        const BitTorrent::TorrentDescriptor &torrentDescr = parseResult.value();
+        const BitTorrent::InfoHash &infoHash = torrentDescr.infoHash();
+
+        const BitTorrent::Torrent *const torrent = BitTorrent::Session::instance()->findTorrent(infoHash);
+        // torrent has already been added and has metadata
+        if (torrent && torrent->info().isValid())
+        {
+            // the TorrentInfo returned by Torrent::info() contains an empty list of trackers
+            // we must fetch the trackers directly via Torrent::trackers()
+            QVector<BitTorrent::TrackerEntry> trackers;
+            for (const BitTorrent::TrackerEntryStatus &tracker : torrent->trackers())
+                trackers << BitTorrent::TrackerEntry
+                {
+                    .url = tracker.url,
+                    .tier = tracker.tier
+                };
+
+            setResult(serializeTorrentMetadata(torrent->info(), trackers));
+            return;
+        }
+
+        const auto iter = m_torrentMetadata.find(infoHash);
+
+        // metadata has already been downloaded
+        if (iter != m_torrentMetadata.end() && iter.value().isValid())
+        {
+            const BitTorrent::TorrentInfo &metadata = iter.value();
+            setResult(serializeTorrentMetadata(metadata, metadata.trackers()));
+            return;
+        }
+
+        if (!session->isKnownTorrent(infoHash) && iter == m_torrentMetadata.end())
+        {
+            qDebug("Fetching metadata for %s", qUtf8Printable(infoHash.toTorrentID().toString()));
+            if (!session->downloadMetadata(torrentDescr))
+                throw APIError(APIErrorType::BadParams, tr("Unable to download metadata"));
+
+            m_torrentMetadata.insert(infoHash, BitTorrent::TorrentInfo{});
+        }
+
+        setResult(QJsonObject {}, 202);
```

Comment:
I wouldn't change setResult but provide corresponding exception for this case to not break current API design.

Reply:
To me it makes logical sense that `setResult` would accept content, a status code, a mime type, etc. What's the downside of "breaking" (i.e. modifying) current API design?

Reply:
> To me it makes logical sense that `setResult` would accept content, a status code, a mime type, etc.

The layer we talked about (implemented as API controller classes) is designed to be HTTP (or any other network protocol) independent. Its area of responsibility is to parse request, forward them to the application core, maintain some intermediate state between requests (if any functions require it), and serialize the results. Another layer is responsible for delivering requests and sending responses via HTTP.
>What's the downside of "breaking" (i.e. modifying) current API design?

Breaking things around you in between is always a bad thing, isn't it? If you have an idea of a better architecture than we have now, and you are ready to implement it or at least outline its concept in sufficient detail so that someone else can implement it, you are welcome.

Reply:
> The layer we talked about (implemented as API controller classes) is designed to be HTTP (or any other network protocol) independent. Its area of responsibility is to parse request, forward them to the application core, maintain some intermediate state between requests (if any functions require it), and serialize the results. Another layer is responsible for delivering requests and sending responses via HTTP.

How would you handle this then? We still need some way of passing an HTTP status code. I can add a `setStatus` method to the `APIController` class but I'm not sure if this would also violate the constraints you've set.

Reply:
> How would you handle this then?

As I suggested initially, through a suitable exception (this is how it is supposed to work if it is impossible to return the result). If you really believe that it should be interpreted in some way other than the regular "not found", you can add a new type to the `APIError` and handle it in `WebApplication`. I guess you can find fault with the fact that it's not an "error" case. These are conventions. It is quite acceptable at this layer to interpret any case of inability to return a result as an "error". HTTP layer should care about transforming it into HTTP-relevant form.

Reply:
Even from an HTTP perspective, I personally don't see much point in returning 202 rather than just 404. The "downloading" torrent metadata process has an "indeterminate" essence. It may never end for unspecified reasons.

Reply:
I think you misunderstand HTTP status codes. I also think you're missing the context that this code is running asynchronously and not in the path of an HTTP request.

Reply:
> I also think you're missing the context that this code is running asynchronously and not in the path of an HTTP request.

As I understand it, it is used for similar cases, i.e. when you give the server a task that it cannot perform (or complete) immediately in order to return the result in current response. I just meant that "searching metadata" differs from a regular task, which has a certain scope of execution. So you can't "promise" the client that "I'll complete this task then and there", you can't provide any progress, etc. The only thing that distinguishes it from the usual "not found" is that you take some action to find the metadata in between. But all the client has from this is that he could try to request it again after a while, which he could do after receiving the 404 code as well.

As per https://restfulapi.net/http-status-202-accepted/:
>When using HTTP 202, it’s important to follow these best practices:
>* Provide a way for the client to monitor the progress or receive the result of the request.
>* Include relevant headers, such as ‘Location‘, to indicate where the client can obtain more information about the request’s status.
>* Clearly document the expected behavior and handling of the request, including any time limits for polling or callbacks.

It doesn't look like we can satisfy the above in this case.

Reply:
Anyway, I don't care much about the HTTP layer. I have expressed my opinion on this, but I leave the rest to your discretion.

Reply:
> I've reverted the changes to `setResult` and added a new `setStatus` method.

No matter how exactly you make this layer dependent on HTTP, it doesn't change anything in terms of my claims.

Reply:
You've not suggested a better alternative, so I'm going to keep it as is. The only suggestion was to throw an error which would not be appropriate.

Reply:
> You've not suggested a better alternative, so I'm going to keep it as is.

The simplest thing to do right now, if you don't want to delay your current PRs, is just to stick to the existing possibilities (i.e., not distinguish between successful statuses, and just return 200 in all such cases). When I (or someone else) improve the API implementation so that it supports various status codes in a way consistent with the existing architecture, you can adjust the returned status codes (IMO, it's better to do this through all APIs at once).

Reply:
This is not a viable solution. The API should do a much better job distinguishing being status codes. Some current APIs even use a 200 to indicate an error. I think the more pragmatic approach is to move forward with this implementation and then adjust the implementation when you/someone has modified the existing architecture to support status codes in some other way. Your approach would require a breaking change in the API when we later change the returned status code.

Reply:
> Your approach would require a breaking change in the API when we later change the returned status code.

Well, this could be considered a valid reason, if we talk only about these new endpoints. But as I mentioned, similar improvements would need to be applied to the entire WebAPI, so it would be a breaking change anyway.

Reply:
> similar improvements would need to be applied to the entire WebAPI,

Otherwise, if the use of these status codes only in new API endpoints and thus making them inconsistent with existing similar endpoints will be considered acceptable, it still does not look difficult to do this without mixing logical layers. All you need is to change "status" inside API controllers layer to be defined in its own (HTTP independent) terms and translate it to HTTP status codes based on them like it's done for translating API errors to HTTP errors. E.g. for the start the corresponding `enum` can consist of `OK` and `Async` items. Then `OK` will be translated to HTTP 200 if endpoint returns some content or to HTTP 204 otherwise, and `Async` will be translated to HTTP 202.

Reply:
> But as I mentioned, similar improvements would need to be applied to the entire WebAPI, so it would be a breaking change anyway.

An approach like this is a great way to ensure that nothing ever gets done because the amount of work required is too monumental. We need iterative improvement.

> All you need is to change "status" inside API controllers layer to be defined in its own (HTTP independent) terms and translate it to HTTP status codes based on them like it's done for translating API errors to HTTP errors. E.g. for the start the corresponding enum can consist of OK and Async items. Then OK will be translated to HTTP 200 if endpoint returns some content or to HTTP 204 otherwise, and Async will be translated to HTTP 202.

Great, if that's "all" that's needed then feel free to put up a PR.

Reply:
> Great, if that's "all" that's needed then feel free to put up a PR.

It's "all" that needs to be changed in this PR to conform to current multilayer design.

Reply:
I've force pushed a change to this PR's initial commit to work off of a `APIStatusType` enum.